### PR TITLE
Set cache control in the HTTP-RPC response header

### LINKED ISSUE #5764: rpc: http jsonrpc don't set cache-control in response
## Summary

Some of the RPC apis can be perfectly cached since the data is immutable, like `/block` `/block_results`, etc.
If we set proper cache-control response header, and setup with some http proxy cache server like [varnish](https://varnish-cache.org/), we can easily archive great performance of http apis.

## Proposal

____

#### For Admin Use

- [x] Not duplicate issue
- [x] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by JayT106**
This PR added the "cache-control" into the HTTP-RPC response header and set the cache expired time to one year

Closes #5764

### CODE REVIEW

File: rpc/jsonrpc/server/http_server.go [hunk 1]
```diff
@@ -146,6 +146,7 @@ func WriteRPCResponseHTTP(w http.ResponseWriter, res ...types.RPCResponse) error
 		return fmt.Errorf("json marshal: %w", err)
 	}
 	w.Header().Set("Content-Type", "application/json")
+	w.Header().Set("Cache-Control", "max-age=31536000") // expired after one year
```
Comment (yihuang) [inline rpc/jsonrpc/server/http_server.go:149]: I think there's some endpoints are not supposed to be cache, for example `/status`, the response will change according to current status.

Comment (alexanderbez) [reply rpc/jsonrpc/server/http_server.go:149]: Yeah I think this is advisable. We should instead consider adding a cache policy on individual requests. Also, one year seems too large, no?

Comment (JayT106) [reply rpc/jsonrpc/server/http_server.go:149]: @alexanderbez Agree with the cache policy that might need to be added. @yihuang, I will check the RPCs and add constrain to them.

**Comment by JayT106** [comment]
@alexanderbez are these routes files not allowed to modify?

**Comment by alexanderbez** [comment]
Not sure I follow @JayT106? What do you mean exactly? Can they override the HTTP header?

**Comment by JayT106** [comment]
> Not sure I follow @JayT106? What do you mean exactly? Can they override the HTTP header?

my bad, I didn't solve the file conflict in the PR which I was thinking about why it happens after I did rebase of the commits.

**Comment by JayT106** [comment]
golangci-lint 2 errors related with the unused code. Can I just remove it?
::error file=light/proxy/routes.go,line=162,col=6::`makeBlockSearchFunc` is unused (deadcode)
::error file=light/proxy/routes.go,line=154,col=6::type `rpcBlockSearchFunc` is unused (unused)

File: light/proxy/routes.go [hunk 2]
```diff
@@ -18,36 +18,35 @@ func RPCRoutes(c *lrpc.Client) map[string]*rpcserver.RPCFunc {
...
-		"net_info":             rpcserver.NewRPCFunc(makeNetInfoFunc(c), ""),
-		"blockchain":           rpcserver.NewRPCFunc(makeBlockchainInfoFunc(c), "minHeight,maxHeight"),
-		"genesis":              rpcserver.NewRPCFunc(makeGenesisFunc(c), ""),
-		"block":                rpcserver.NewRPCFunc(makeBlockFunc(c), "height"),
-		"block_by_hash":        rpcserver.NewRPCFunc(makeBlockByHashFunc(c), "hash"),
-		"block_results":        rpcserver.NewRPCFunc(makeBlockResultsFunc(c), "height"),
-		"commit":               rpcserver.NewRPCFunc(makeCommitFunc(c), "height"),
-		"tx":                   rpcserver.NewRPCFunc(makeTxFunc(c), "hash,prove"),
-		"tx_search":            rpcserver.NewRPCFunc(makeTxSearchFunc(c), "query,prove,page,per_page,order_by"),
-		"block_search":         rpcserver.NewRPCFunc(makeBlockSearchFunc(c), "query,page,per_page,order_by"),
```
Comment (cmwaters) [inline light/proxy/routes.go:32]: You seem to have omitted `block_search`

Comment (JayT106) [reply light/proxy/routes.go:32]: ah, gotcha. Thanks

File: rpc/jsonrpc/server/http_server.go (hunk 1 above)
Comment (JayT106) [reply rpc/jsonrpc/server/http_server.go]: @alexanderbez Was my intuition, however, found a link to explain why set the age to one year if we see the data is immutable (but varnish cache doesn't support it)
https://ashton.codes/set-cache-control-max-age-1-year/

Comment (alexanderbez) [reply rpc/jsonrpc/server/http_server.go]: Absolutely. If the response data does not change, then this makes sense 👍

File: rpc/jsonrpc/server/http_json_handler.go [hunk 3]
```diff
@@ -57,6 +57,9 @@ func makeJSONRPCHandler(funcMap map[string]*RPCFunc, logger log.Logger) http.Han
 			requests = []types.RPCRequest{request}
 		}
 
+		// Set the default response cache to true unless we see any error and any rpc request doesn't
+		// allow to be cached
+		var c = true
```
Comment (cmwaters) [inline rpc/jsonrpc/server/http_json_handler.go:62]: There is one other condition that we have to be wary about. For a lot of RPC requests i.e validators, consensus params, blocks and commits, if a client requests this data with a height of 0 the node responds by sending that data type at the latest height. This means that the same request of height 0 to these RPC endpoints should also over time send different responses and thus shouldn't cache the response

Comment (alexanderbez) [reply rpc/jsonrpc/server/http_json_handler.go:62]: That is correct!

Comment (JayT106) [reply rpc/jsonrpc/server/http_json_handler.go:62]: Thanks for the feedback, I will check the condition and revise the logic.

Comment (cmwaters) [reply rpc/jsonrpc/server/http_json_handler.go:62]: Awesome. Once this is done I think the PR should be ready to merge

Comment (JayT106) [reply rpc/jsonrpc/server/http_json_handler.go]: Added default height checking for these types of RPC requests.

**Review by cmwaters — APPROVED** [review]
Awesome work @JayT106 🚀

**Pull request merged**